### ISSUE TITLE
Case with EmailMessage and LiveChatTranscript - Parent Trigger Errors with Scripts

### ISSUE THREAD

**Issue by rachel724** [SFDO-Community/declarative-lookup-rollup-summaries#1038]
**Describe the bug**
In creating a Rollup from EmailMessage and LiveChatTranscript to Case, I am getting errors that my Relationship Criteria fields are not available on the Case object when a script is run against Cases.  I do not get the error when I activate or run a full calculate, nor do I get the error throughout the day when the lookup is running.

**To Reproduce**
1. Create a Rollup with the following information:
Parent Object: Case
Relationship Field: ParentId
Child Object: Email
Relationship Criteria: Incoming = FALSE
Relationship Criteria Fields: Incoming
Field to Aggregate: CreatedDate
Aggregate Option: Last
Field to Aggregate To: (Any custom date/time field on Case)
Calculation Mode: Realtime
Calculation Sharing Mode: System

2. Save
3. Deploy Child Trigger
4. Activate and Save Rollup
5. Run Full Calculate
6. Run a script to update any field on Case

**Expected behavior**
I would expect an update at the case level to not run into any issues with a dlrs on a child object.  The script should run successfully.

**Screenshots**
![Screen Shot 2021-06-03 at 12 26 47 AM](https://user-images.githubusercontent.com/13396949/120591890-8711c700-c402-11eb-8bfa-279549a2adb2.png)

**Additional context**
I have tried removing all of the triggers and deleting the dlrs records and building and deploying again, and have narrowed it down to Case with EmailMessage and LiveChatTranscript.  It is acting as if I am saying that the fields are on the Case object.  It is also different per dlrs.  It sometimes is the relationship criteria field that shows up in the error, sometimes the field to aggregate.  It seems to be field to aggregate first, unless that field exists on Case (like CreatedDate), and if that field exists on Case, it then picks the relationship criteria field.

**Comment by aheber** [comment]
@rachel724 are you still experiencing the problem? Would love to try and help figure out what is going on and get this resolved.

Could you provide a screenshot of the Rollup configuration screen?

**Comment by bdsdx** [comment]
@schmalli I came to very similar conclusion (detailed in #1354). One point, `Incoming` does not exist on the Case object. But it does on the EmailMessage one.

**Comment by schmalli** [comment]
@bdsdx Yes sure, this was confusing 🙂 Sorry, I wanted to refer to the error message 
> Invalid field Incoming for Case

In general the issue is always referring to fields not existing on Case.
Thanks for clarification! 🙌

**Comment by aheber** [comment]
For those experiencing this issue, can you report your version?

There are some similar problems fixed in v2.16 (released a couple of years ago) that should have helped here and I'm trying to track down if this is a problem in the recent versions or if this is resolved now.

Thanks.

**Comment by aheber** [comment]
@SFSod I'm currently unable to make progress on it because I haven't been able to get clean-room reproduction steps.

If you think you could work on that, I would need clear steps to take in a fresh org that would reproduce this problem. Unfortunately so far I've only had people that can reproduce it in a Sandbox but nobody has been able to make it happen in a clean org and provide me steps.

Good to know it is still a problem in the current version. As soon as I can figure out how to isolate and replicate then we can work on getting it fixed.

**Comment by jack071900** [comment]
Is there an update on this bug. We are having this problem with the Case object when users try to merge. We are looking for a solution.

**Comment by aheber** [comment]
@jack071900 the development team hasn't been able to replicate this in a clean org. That prevents us from finding a root cause.

I was working with someone else to replicate it this morning, I'm hoping that effort will bear fruit in the next week or two but I can't promise anything.

The most useful thing you could do to move this along would be to figure out how to replicate this behavior in a clean setup. There is a setting/configuration/toggle/etc... that is different in the orgs having trouble and I don't know what it is. I'm sorry.

Hopefully we can figure it out shortly.

**Comment by jack071900** [comment]
@aheber We find that it is impacting some records but not all. We think the records that are impacted are created with email-to-case, but I haven't confirmed that yet. We are going to try to dlrs_CaseTrigger so it does not trigger when status = Merged.

**Comment by aheber** [comment]
@maik-idealo maybe you could add your modified Case Trigger here?

Just keep in mind that this means that merged cases will hold incorrect rollup values. You'll probably wan't to use Schedule Full Calculate to true up the records once a day or once a week.

If you don't have rollups where the Case is the child then you could use the Manage Child Trigger button and remove the trigger on Case. It is there if Case is the child object of a rollup or only if Case is a parent and you've chosen to add the trigger to monitor for merges. When managing the trigger where Case is a parent you could change the checkbox that asks if you want to add the trigger to the parent to support merging. That is only a good idea if you don't need it where Case is a child and you're willing to let the merged cases fall out of sync for a time.

**Comment by jack071900** [comment]
@aheber Case is the child object.

Is there a modified Case Trigger?

**Comment by jack071900** [comment]
No, we don’t have any dlrs where case is the parent object. We only have one dlrs which we added in Dec 2024 where case is the child object.

**Comment by aheber** [comment]
OK, very interesting. What is the specific error you are getting? Could you also share your rollup configuration?

Do you have other rollup configured? Where Case is the child, what is the Parent? Are there other rollups where that same object is parent or child?

**Comment by jack071900** [comment]
We use a lot of DLRS and we have a lot of very positive experience with DLRS.

We only have one DLRS that involves CASE and CASE is the child. Most of our DLRS are scheduled, but this one is real time. When I update this DRLS with “Status != Merge" it doesn’t solve the problem. When I make the DLRS inactive, it doesn’t fix the problem. When I turn off the trigger, the problem is resolved.

We agree with you, most of our cases are used by our Support team. They merge cases and there are no problems. But our internal sales ops team can’t merge cases. We get a flow error. We think that most of their cases are created by email-to-case, but I need to validate that.

Here is the config for the one DLRS where Case is child:

![image](https://github.com/user-attachments/assets/3eb31d93-9938-4d39-a9ca-6a4162a79388)

Here is the error message: 
![image](https://github.com/user-attachments/assets/fadf938f-fdbe-4004-822f-35361563e567)

When I try to replicate it in one of our sandboxes, I get an additional problem. Cases where status = Merged cannot be edited or they cause that error. If I turn off that flow (or add start criteria "status != merged") then another flow causes the same error. We have a lot of flows on the case object so that doesn't seem to be a viable path.

Thanks for your help.

**Comment by aheber** [comment]
@jack071900 thanks for that info. Could you try and capture a debug log during that execution? I'm hoping a more clear error is present there.

In your rollup config, can you tell me what you have for Relationship Criteria Fields?

Can you show me the content of the Case Trigger that you have in place now?

**Comment by jack071900** [comment]
Relationship Criteria Fields: 
Status

dlrs_CaseTrigger:

/** * Auto Generated and Deployed by the Declarative Lookup Rollup Summaries Tool package (dlrs) **/trigger dlrs_CaseTrigger on Case    (before delete, before insert, before update, after delete, after insert, after undelete, after update){    dlrs.RollupService.triggerHandler(Case.SObjectType);

debug log attached

[dlrs debug for aheber.xlsx](https://github.com/user-attachments/files/18446028/dlrs.debug.for.aheber.xlsx)

Given that most of our cases can merge without a problem, it is not an org config. Given that the problem is primarily with one record type, but not exclusively, it is not likely an issue with record type configuration. It must be with a field or value or something else that is record specific.

**Comment by aheber** [comment]
@jack071900 thank you for that. The debug log indeed shows a little more detail. Something is confusing DLRS and it is trying to read a field of `Name` from `Case` which doesn't have that, it uses `CaseNumber` in place of name.

Could you look at your rollup configurations and see which, if any, have `Name` in the Field to Aggregate, Relationship Field, or Relationship Criteria Fields? (I think that is all the places I would expect this to pull from)

Hoping knowing which rollup this is getting confused with would help us to find the connection.

It only happening sometimes doesn't exclude environmental causes, only that something about those records is needed to help cause the behavior. Any additional information you can gather about those problematic records, and any tests you can do to isolate the cause, would be incredibly helpful.

If you'd like to open a support case and request that they capture debug logs while you replicate the error that would be helpful. They may push back because this is a managed package. If so let me know and we'll get someone in Salesforce to approve the action.

**Comment by jack071900** [comment]
@aheber 
 We use "Name" in two other DLRS. One is Assets (child) to Accounts (parent). The other one is two custom objects.

We've seen this error with other fields as well, such as "Assigned_to_Profile__c". This field is used on three DLRS.

I turn off all dlrs with the field name in the error message and then tried to edit the record that displays the problem. I get the same error but with a different field. I look for that field in the dlrs, turn them off, and try again. And I get a different field name in the error message.

one of the field names in the error message is used in the relationship criteria.
two of the field names in the error message are in the field to aggregate.

I am checking on whether I can send you a full log. I feel confident I can convince SF to do the log, but I may need you to sign a doc before I can send you a full log.

**Comment by jack071900** [comment]
@aheber In one of my sandboxes, I have a record that cannot be edited because of the dlrs error. When I clone the record so the status is no longer "Merged" and the record no longer has a Master Record ID, then there is no error.

**Comment by aheber** [comment]
@jack071900 thank you for those updates. I've been thinking about them for a few days and haven't figured out what would cause it yet.

For the cloned record, does it clone the child records? I suspect there is some correlation with related records hanging off of it and confusing it. I haven't been able to figure out what it is yet.

I really appreciate the testing you're doing. As much as you can rule things out and get that sandbox down to the most limited config would be helpful to building a replication scenario that I can run on my side.

**Comment by aheber** [comment]
For all those following along. I have a replication of this problem. It is somewhat abstract, I still can't figure out how to get the platform to do it during normal use but I can force it to happen using test code that is carefully configured. With what I'm seeing it should match closely enough to real world scenarios that I believe I've narrowed down to the root cause.

In certain circumstances the rollup code that checks if a merge is happening right now is tripping over itself and processing the merged item against rollup configurations where it is both the parent and child. The errors comes when the merged item is the parent of the configuration but DLRS is checking it using the Child fields which don't exist on the parent.

Handling merging can be complicated and, as seen when troubleshooting, can take a variety of different of different execution paths. I'll have to build our catalog of test cases before I'm comfortable tackling this fix.

I don't think the code change is huge, the hard part will be making sure I'm confident in the change. Once the fix is ready for testing I'll ask this group to test a beta package in a Sandbox to see if it resolves the problem.

**Comment by jack071900** [comment]
@aheber Thanks for the update, this is encouraging. We are hitting a lot of problems because of this. I appreciate you help. And I am encouraged that you seem to have figured some of it out. Let me know how I can help.

**Comment by aheber** [comment]
@jack071900 on the records you are having a problem with, can you look at the `MasterRecordId` field? I have a theory that the problem records might already have that populated and that is part of what is causing the problems. I could be wrong still.

**Comment by jack071900** [comment]
@aheber No, they don't have master record ids.

**Comment by aheber** [comment]
@jack071900 thank you for checking. That kills one theory... more testing.

It shouldn't hamper the expected fix though, just one thought about why some records are running into this.

**Comment by jack071900** [comment]
@aheber After further research, I think it is related to the Master Record ID. 

Here is the scenario where we experience this problem.

1. There is a DLRS on the Case object.
2. There is one or more DLRS on objects related to the Case object.
4. There is one or more flow on objects related to the Case object.
5. The Merge Case settings are configured to not delete the losing record.
6. The post-merge losing record has a Master Record ID.

I wonder if this would be resolved if the dlrs trigger omitted case records that have a master record id. This is an archived record. It should not be updated because another record has been identified as the winning version.

**Comment by aheber** [comment]
I'm still digging. What I'm finding is that the merge functionality was built before you could merge cases and especially before you could retain cases in a merge. This means the merge logic would generally only see a record having a `MasterRecordId` when it was being deleted and all the children reparented to the surviving merge target.

With the ability to keep merged cases around we can see a previously merged case still get edits, this confuses the DLRS merge logic because part of that process is find all the rollups on child objects and cause them to be recalculated if they care about the merged parent.

DLRS grabs not only the rollups for the Case object but also any rollups for object which have lookups to the Case object (because those items might need to be recalculated because they were reparented). Then it tries to pull any of those fields from the Case object, at which point we see the error about the Case not having that field.

I think there are variations of the case/merged object showing up in the triggers with the MasterRecordId during the merge cycle or after. It seems to all be the same logic problem though. DLRS isn't expecting to see an edited record in the triggers while it has a `MasterRecordId` so when it runs the edit checks it gets confused. I can't make it happen naturally but I can force it into these situations and that should generally be good enough to fix it.

This is all pretty narrow in terms of code. I just don't have the guard rails in place right now to make sure merging isn't damaged if I fix this logic. I have to build the guard rails (Apex test code) first then I can cleanup the merge instructions and hopefully make some much needed improvements along the way.

**Comment by jack071900** [comment]
@aheber Thank you, we appreciate your work to fix this.

**Comment by aheber** [comment]
OK, testing time! I think we have the code settled to fix this problem :tada: I really appreciate so many of you waiting in the wings while we figured out what was causing this. Certainly one of the more difficult ones to track down.

Now I need people's help! Before releasing this next version I would appreciate if people would test the beta package to confirm the problem is resolved. (make sure you have a "broken" record on hand, try and merge with the existing v2.22, instal the beta package and try merging again)

A few things to keep in mind. Beta packages can only be installed in non-production orgs. (Sandboxes are probably best)

Once you install a beta package in your sandbox you can't upgrade it. You will either have to refresh that sandbox or uninstall and reinstall a non-beta version of the package. Only do this testing in an org that you're OK with it getting a beta package and possibly not being able to upgrade it without a refresh.

If you would be willing to confirm this problem is resolve you can use this install link: https://test.salesforce.com/packaging/installPackage.apexp?p0=04tKA000000cBO1YAM
(Relase details: https://github.com/SFDO-Community/declarative-lookup-rollup-summaries/releases)

My goal is for the next couple of weeks to be testing and to have this version released mid-March so as long as some of you can confirm this fix then your prod orgs could be fixed to go as soon as that.

**Comment by maik-idealo** [comment]
Hi @aheber,

Thank you. I just tested the new beta. The error is solved, but the recalculation after the merge is not working.

Emails of Case test 1 before merge: 0
Emails of Case test 2 before merge: 1

After merge of Case test 2 into Case test 1:

The DLRS calculation was not executed and the field for the number of Emails is still 0

**Comment by aheber** [comment]
@maik-idealo can you check your Scheduled Items? Merges run into problems if they try and rollup right away so they will always create a scheduled item even if they are in realtime mode.

If you look in the Scheduled Items you should see the queued up record and you should be able to run the RollupJob class to process it and get the correct value.

Let me know if you've already done all that and it still didn't work.

Appreciate you testing it, hopefully that is all it is.

**Comment by maik-idealo** [comment]
Hi @aheber 
That's right, the items were there. After I activated a schedules apex job and it was executed, the recalculation was performed.
So the issue looks solved according to my tests. Thank you very much, good work!

**Comment by aheber** [comment]
Great news!! Thank you so much for putting it through the paces. Hopefully we can get a few more confirmations before we finalize this build. This is a huge confirmation!

**Comment by aheber** [comment]
This change will ship in the upcoming v2.23, should be released next March. Thanks again to everyone who contributed to finding the root cause.

**Comment by marcinmietusrad** [comment]
@aheber do you know exact date for release 2.23? So I can deploy this to Prod?

**Comment by aheber** [comment]
@marcinmietusrad we are still completing release actions but the package is available now.
https://github.com/SFDO-Community/declarative-lookup-rollup-summaries/releases/tag/release%2F2.23

Direct install link
https://login.salesforce.com/packaging/installPackage.apexp?p0=04tKA000000cBQWYA2
